### ISSUE TITLE
Regression: Sum type doesn't raise exception on invalid input

### ISSUE THREAD

**Issue by Kukunin** [dry-rb/dry-types#164]
Hopefully this chunk of code describes a problem

```ruby
    class A < Dry::Struct
      attribute :attr, Types::Strict::String
    end

    C = A | A
    A[other: ''] # raises Dry::Types::ConstraintError in dry-types 0.9.0 and 0.9.1
    C[other: ''] # raises Dry::Types::ConstraintError in dry-types 0.9.0 but returns {other: ''} in 0.9.1
```

**Comment by solnic** [comment]
Ah damn, we should make structs constrained types. @flash-gordon wdyt about making `Struct.constrained?` return `true`? I'd consider all structs as constrained types

**Comment by solnic** [comment]
This has been fixed in dry-struct v0.1.1

**Comment by flash-gordon** [comment]
@solnic I had missed the mail, then found the bug myself >.<
